core/src/main/java/hudson/util/RobustCollectionConverter.java

```diff
@@ -52,14 +56,33 @@
 @SuppressWarnings({"rawtypes", "unchecked"})
 public class RobustCollectionConverter extends CollectionConverter {
     private final SerializableConverter sc;
+    /**
+     * When available, this field holds the declared type of the collection being deserialized.
+     * This is used to ensure that deserialized elements have the expected type.
+     */
+    // TODO: We only support one level of parameterization, i.e. the elements of a List<List<Integer>>
+    // will be treated as List<Object>, and so the inner list could have invalid non-Integer elements. It is unclear if
+    // we can do better without significant changes.
+    private final @CheckForNull Class<?> elementType;
 
     public RobustCollectionConverter(XStream xs) {
-        this(xs.getMapper(), xs.getReflectionProvider());
+        this(xs.getMapper(), xs.getReflectionProvider(), null);
     }
 
     public RobustCollectionConverter(Mapper mapper, ReflectionProvider reflectionProvider) {
+        this(mapper, reflectionProvider, null);
+    }
+
+    public RobustCollectionConverter(Mapper mapper, ReflectionProvider reflectionProvider, Type collectionType) {
         super(mapper);
         sc = new SerializableConverter(mapper, reflectionProvider, new ClassLoaderReference(null));
+        if (collectionType != null && Collection.class.isAssignableFrom(Types.erasure(collectionType))) {
+            var baseType = Types.getBaseClass(collectionType, Collection.class);
+            var typeArg = Types.getTypeArgument(baseType, 0, Object.class);
```

Comment:
What happens here if the field is using a rawtype?

Reply:
Yes good point, I will add a test case for rawtypes.

Reply:
If we were using `Types.getTypeArgument(Type, int)`, rawtypes would cause an `IllegalArgumentException`. With `Types.getTypeArgument(Type, int, Type)` the default value just gets used instead. I'll still add test cases though.

Reply:
Tests added in [312e51c](https://github.com/jenkinsci/jenkins/pull/9727/commits/312e51c9464c8f6d2342a7a0a7983cab2f00b5b9).